### DISCUSSION TITLE
JSON Lines

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#19] [Ideas]
Data management in JSON lines seems to be a promising approach:
* easy to parse, because each line is formatted in JSON
* memory efficient, because each line can be processed independently

**Comment by nichtich** [comment]
JSON lines is known [under mutliple names](https://en.wikipedia.org/wiki/JSON_streaming#Newline-delimited_JSON) (NDJSON, ND-JSON, JSON-ND, JSONL...). PG data in one line per node or edge could be defined as

>  **PG-NDJSON** is a serialization of the property graph data model in newline-delimited JSON (NDJSON). A       graph is encoded as sequence of lines with a JSON object encoding either a node or an edge, respectively. The format of these objects is identical to the corresponding node or edge objects in PG-JSON.

Note that some Neo4J can export to line based JSON too, see https://neo4j.com/labs/apoc/4.1/export/json/ with option `JSON_LINES`. The format uses different field names than PG-JSON / PG-NDJSON:

~~~
{"type":"node","id":"0","labels":["User"],"properties":{"born":"2015-07-04T19:32:24","name":"Adam","place":{"crs":"wgs-84","latitude":13.1,"longitude":33.46789,"height":null},"age":42,"male":true,"kids":["Sam","Anna","Grace"]}}
{"type":"node","id":"1","labels":["User"],"properties":{"name":"Jim","age":42}}
{"type":"node","id":"2","labels":["User"],"properties":{"age":12}}
{"id":"0","type":"relationship","label":"KNOWS","properties":{"since":1993,"bffSince":"P5M1DT12H"},"start":{"id":"0","labels":["User"]},"end":{"id":"1","labels":["User"]}}
~~~

This can easily be converted to and from PG-NDJSON, expect for custom datatypes (here: coordinate `{"crs":"wgs-84","latitude":13.1,"longitude":33.46789,"height":null}`).

**Comment by hchiba1** [comment]
Thanks for the information.
Do you prefer the name "NDJSON"?
Actually, I was not familiar with it, and just looked it up.
https://github.com/ndjson/ndjson-spec/issues/35#issuecomment-1285673417
I still have the impression that JSON lines is a bit more popular around me.
Currently, I like PG-JSONL

**Comment by nichtich** [comment]
I think there is no consensus. We always use `.ndjson` and I've seen it in several other projects as well. Other communities use `.jsonl`.

> Currently, I like PG-JSONL

This would make two formats named PG-JSON and PG-JSONL, right?

**Reply by hchiba1** [reply]
Yes. Perhaps polls may also help the discussion.

**Reply by nichtich** [reply]
Depends on who you ask :-) Searching on GitHub for file extension shows that `.jsonl` is used more, so I'll rename it in pgraph implementation, ok? `ndjson` may still be accepted as alias.

**Comment by hchiba1** [comment]
When I format in PG-JSONL, I would insert space after `:` and `,`
(ref. https://jsonlines.org/examples/).

**Comment by hchiba1** [comment]
Currently, a simple PG `a` is converted to PG-JSONL as follows:
```
{"type": "node", "id": "a", "labels": [], "properties": {}}
```
Should we explicitly write empty arrays and objects?

**Reply by nichtich** [reply]
Please see https://github.com/pg-format/specification/blob/main/README.md#pg-jsonl for the draft of a formal specification: empty arrays and objects MUST be written but applications SHOULD accept lines where these fields are omitted.
